src/Symfony/Component/DependencyInjection/ServiceSubscriberTrait.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection;
+
+use Psr\Container\ContainerInterface;
+
+/**
+ * Implementation of ServiceSubscriberInterface that determines subscribed services from
+ * private method return types. Service ids are available as "ClassName::methodName".
+ *
+ * @author Kevin Bond <kevinbond@gmail.com>
+ */
+trait ServiceSubscriberTrait
+{
+    /** @var ContainerInterface */
+    private $container;
+
+    public static function getSubscribedServices(): array
+    {
+        static $services;
+
+        if (null !== $services) {
+            return $services;
+        }
+
+        $services = \is_callable(array('parent', __FUNCTION__)) ? parent::getSubscribedServices() : array();
+
+        foreach ((new \ReflectionClass(self::class))->getMethods() as $method) {
+            if ($method->isStatic() || $method->isAbstract() || $method->isGenerator() || $method->isInternal() || $method->getNumberOfRequiredParameters()) {
+                continue;
+            }
+
+            if (($returnType = $method->getReturnType()) && !$returnType->isBuiltin()) {
```

Comment:
I feel like we should be even more strict an not wire methods from parents, as they should be the one declaring their deps.
$method->getDeclaringClass() === self::class should be added first in the "if" (then we can use self::class also on the next line.)